---
title: "M17_check"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a 'pipeline' to check whether or not samples contain the M17 gene.

```{r}
#get files
#system("nohup bash /home/$USER/scpackage/data-raw/get_data.sh")

#assemble fastq's
#system("nohup bash /home/$USER/scpackage/inst/megahit.sh")

#get contigs in one fasta
add_accession_contigs_megahit(samples_dir = "/home/$USER/research_drive/geodescent/samples/MGYS00005036/",
                              sra_sampletype = "SRR",
                              sample_nums_from = 6231130,
                              sample_nums_to = 6231221,
                              outfile = "/home/$USER/research_drive/geodescent/samples/MGYS00005036/blast/contigs/all_samples_urban1.fa")

#create blast db
makeblastdb(input = "/home/$USER/research_drive/geodescent/samples/MGYS00005036/blast/contigs/all_samples_urban1.fa", 
            outname = "megahit_untrimmed_urban1", 
            outdir = "/home/$USER/research_drive/geodescent/samples/MGYS00005036/blast/",
            dbtype = "nucl")


#run blast against db
blast(blast = "tblastn",
      blast_db = "/home/$USER/research_drive/geodescent/samples/MGYS00005036/blast/megahit_untrimmed_urban1/megahit_untrimmed_urban1",
      input = "/home/$USER/scpackage/inst/extdata/M17.txt",
      out = "/home/$USER/research_drive/geodescent/samples/MGYS00005036/blast/test_tblastn_urban1_2",
      format = 6)

```


