---
title: "M17_check"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This following code block is to quickly check if a wgs dataset contains the marker genes supplied.

```{r}
#get files
#system("nohup bash /home/$USER/scpackage/data-raw/get_data.sh")

#assemble fastq's
#system("nohup bash /home/$USER/scpackage/inst/megahit.sh")

#get contigs in one fasta
add_accession_contigs_megahit(samples_dir = "/home/$USER/data/geodescent/samples/MGYS00000991/",
                              sra_sampletype = "ERR",
                              sample_nums_from = 1424899,
                              sample_nums_to = 1424903,
                              outfile = "/home/$USER/data/geodescent/samples/MGYS00000991/blast/contigs/arctic1.fa")

#create blast db
makeblastdb(input = "/home/$USER/data/geodescent/samples/MGYS00000991/blast/contigs/arctic1.fa", 
            outname = "megahit_untrimmed_1", 
            outdir = "/home/$USER/data/geodescent/samples/MGYS00000991/blast/",
            dbtype = "nucl")


#run blast against db
blast(blast = "tblastn",
      blast_db = "/home/rstudio/data/geodescent/samples/MGYS00000991/blast/megahit_untrimmed_1/megahit_untrimmed_1",
      input = "/home/rstudio/scpackage/inst/extdata/M17.txt",
      out = "/home/rstudio/data/geodescent/samples/MGYS00000991/blast/test_tblastn_arctic1a",
      format = 6)

blast(blast = "tblastn",
      blast_db = "/home/rstudio/data/geodescent/samples/MGYS00000991/blast/megahit_untrimmed_1/megahit_untrimmed_1",
      input = "/home/rstudio/scpackage/inst/extdata/gldiB.txt",
      out = "/home/rstudio/data/geodescent/samples/MGYS00000991/blast/test_tblastn_arctic1b",
      format = 6)
```

#Loading in the data

We begin by loading in the filereports of the studies of interest. The url of a filereport of a study can be gotten from the corresponding ebi metagenomic website by saving the link address of the 'TEXT' button (e.g. see https://www.ebi.ac.uk/ena/data/view/PRJEB8968)
```{r}
#getting filereports
#they are saved as rda files in inst/extdata named filereport_accession

#MGYS00000974 ocean cruise
get_filereport(url = "https://www.ebi.ac.uk/ena/data/warehouse/filereport?accession=PRJEB8968&result=read_run&fields=study_accession,sample_accession,secondary_sample_accession,experiment_accession,run_accession,tax_id,scientific_name,instrument_model,library_layout,fastq_ftp,fastq_galaxy,submitted_ftp,submitted_galaxy,sra_ftp,sra_galaxy,cram_index_ftp,cram_index_galaxy&download=txt",
               acc = "PRJEB8968")

#MGYS00005036 urban
get_filereport(url = "https://www.ebi.ac.uk/ena/data/warehouse/filereport?accession=PRJEB26733&result=analysis&fields=analysis_accession,study_accession,sample_accession,secondary_sample_accession,tax_id,scientific_name,submitted_ftp,submitted_galaxy&download=txt",
               acc = "PRJEB26733")

#MGYS00005036 urban
get_filereport(url = "https://www.ebi.ac.uk/ena/data/warehouse/filereport?accession=PRJNA415974&result=read_run&fields=study_accession,sample_accession,secondary_sample_accession,experiment_accession,run_accession,tax_id,scientific_name,instrument_model,library_layout,fastq_ftp,fastq_galaxy,submitted_ftp,submitted_galaxy,sra_ftp,sra_galaxy,cram_index_ftp,cram_index_galaxy&download=txt",
               acc = "PRJNA415974")

#MGYS00000991 arctic
get_filereport(url = "https://www.ebi.ac.uk/ena/data/warehouse/filereport?accession=PRJEB14154&result=read_run&fields=study_accession,sample_accession,secondary_sample_accession,experiment_accession,run_accession,tax_id,scientific_name,instrument_model,library_layout,fastq_ftp,fastq_galaxy,submitted_ftp,submitted_galaxy,sra_ftp,sra_galaxy,cram_index_ftp,cram_index_galaxy&download=txt",
               acc = "PRJEB14154")

#test
get_filereport(url = "https://www.ebi.ac.uk/ena/data/warehouse/filereport?accession=PRJEB4331&result=read_run&fields=study_accession,sample_accession,secondary_sample_accession,experiment_accession,run_accession,tax_id,scientific_name,instrument_model,library_layout,fastq_bytes,fastq_ftp,fastq_galaxy,submitted_ftp,submitted_galaxy,sra_ftp,sra_galaxy,cram_index_ftp,cram_index_galaxy&download=txt",
               acc = "PRJEB4331")
```


Next we need to get the fastq files of the studies. This is done using the filereport content so it is important to load those in first.
```{r}
#loading in filereports
load("extdata/filereport_PRJNA415974")
f1 = filereport
rm(filereport)
load("extdata/filereport_PRJEB26733")
f2 = filereport
rm(filereport)

load("extdata/filereport_PRJEB4331")

#getting fastq files
get_fastq(filereport = filereport,
          outdir = "/home/rstudio/data/geodescent/samples/TEST")

```

```{r}
#getting metadata tables
get_MGYS00000974()
get_MGYS00000991()
get_MGYS00005036()
```
